R/module_teal_lockfile.R

```diff
@@ -0,0 +1,195 @@
+#' Generate lockfile for application's environment reproducibility
+#'
+#' @param lockfile_path (`character`) path to the lockfile (`"teal_app.lock"`).
+#'
+#' @section Different ways of creating lockfile:
+#' `teal` leverages [renv::snapshot()], which offers multiple methods for lockfile creation.
+#'
+#' - User-specified:
+#'     - **Pre-computed lockfile**: Users can provide their own pre-computed lockfile by specifying the path via
+#'     `teal.renv.lockfile` option. Automatic lockfile computation is skipped in such case.
+#' - Automatically computed:
+#'     - **Working directory lockfile**: If `teal.renv.lockfile` is not set, `teal` will, by default, create an
+#'      `implicit` type lockfile that uses `renv::dependencies()` to detect all R packages in the current project's
+#'      working directory.
+#'     - **`DESCRIPTION`-based lockfile**: To generate a lockfile based on a `DESCRIPTION` file in your working
+#'     directory, set `renv::settings$snapshot.type("explicit")`. The naming convention for `type` follows
+#'     `renv::snapshot()`. For the `"explicit"` type, refer to `renv::settings$package.dependency.fields()` for the
+#'     `DESCRIPTION` fields included in the lockfile.
+#'     - **Custom files-based lockfile**: To specify custom files as the basis for the lockfile, set
+#'     `renv::settings$snapshot.type("custom")` and configure the `renv.snapshot.filter` option.
+#'
+#' @section lockfile usage:
+#' After creating the lockfile, you can restore the application environment using `renv::restore()`.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()].
+#'
+#' @return `NULL`
+#'
+#' @name module_teal_lockfile
+#' @rdname module_teal_lockfile
+#'
+#' @keywords internal
+NULL
+
+#' @rdname module_teal_lockfile
+ui_teal_lockfile <- function(id) {
+  if (!isTRUE(getOption("teal.renv.enable"))) {
+    return(NULL)
+  } else if (!.is_lockfile_deps_installed()) {
+    warning("Lockfile feature disabled. `mirai` and `renv` packages must be installed.")
+    return(NULL)
+  }
+
+  ns <- NS(id)
+  shiny::tagList(
+    tags$span("", id = ns("lockFileStatus")),
+    shinyjs::disabled(downloadLink(ns("lockFileLink"), "Download lockfile"))
+  )
+}
+
+#' @rdname module_teal_lockfile
+srv_teal_lockfile <- function(id) {
+  if (!isTRUE(getOption("teal.renv.enable")) || !.is_lockfile_deps_installed()) {
+    return(NULL)
```

Comment:
it would be good to put something in the logs
we have a few exit criteria and it would be good if we can identify which one occured in case of issues

Reply:
I'll move UI to renderUI and keep only ifs only on the server side 👍